File: examples/go/inventory/logic/state.go
package logic

import (
	angzarrpb "angzarr/proto/angzarr"
	"inventory/proto/examples"
)

// RebuildState reconstructs inventory state from an event book.
func RebuildState(eventBook *angzarrpb.EventBook) InventoryState {
	state := InventoryState{
		Reservations: make(map[string]int32),
	}

	if eventBook == nil || len(eventBook.Pages) == 0 {
		return state
	}

	if eventBook.Snapshot != nil && eventBook.Snapshot.State != nil {
		var snapState examples.InventoryState
		if err := eventBook.Snapshot.State.UnmarshalTo(&snapState); err == nil {
			state.ProductID = snapState.ProductId
			state.OnHand = snapState.OnHand
			state.Reserved = snapState.Reserved
			state.LowStockThreshold = snapState.LowStockThreshold
			for k, v := range snapState.Reservations {
				state.Reservations[k] = v
			}
		}
	}

	for _, page := range eventBook.Pages {
		if page.Event == nil {
			continue
		}

		switch {
		case page.Event.MessageIs(&examples.StockInitialized{}):
			var event examples.StockInitialized
			if err := page.Event.UnmarshalTo(&event); err == nil {
				state.ProductID = event.ProductId
				state.OnHand = event.Quantity
				state.LowStockThreshold = event.LowStockThreshold
			}

		case page.Event.MessageIs(&examples.StockReceived{}):
			var event examples.StockReceived
			if err := page.Event.UnmarshalTo(&event); err == nil {
				state.OnHand = event.NewOnHand
			}

		case page.Event.MessageIs(&examples.StockReserved{}):
			var event examples.StockReserved
			if err := page.Event.UnmarshalTo(&event); err == nil {
				state.Reserved += event.Quantity
				state.Reservations[event.OrderId] = event.Quantity
			}

		case page.Event.MessageIs(&examples.ReservationReleased{}):
			var event examples.ReservationReleased
			if err := page.Event.UnmarshalTo(&event); err == nil {
				qty := state.Reservations[event.OrderId]
				state.Reserved -= qty
				delete(state.Reservations, event.OrderId)
			}

		case page.Event.MessageIs(&examples.ReservationCommitted{}):
			var event examples.ReservationCommitted
			if err := page.Event.UnmarshalTo(&event); err == nil {
				qty := state.Reservations[event.OrderId]
				state.OnHand -= qty
				state.Reserved -= qty
				delete(state.Reservations, event.OrderId)
			}
		}
	}

	return state
}

File: examples/python/inventory/handlers/state.py
"""Inventory state rebuilding logic."""

from dataclasses import dataclass, field

from angzarr import types_pb2 as types
from proto import domains_pb2 as domains


@dataclass
class InventoryState:
    product_id: str = ""
    on_hand: int = 0
    reserved: int = 0
    low_stock_threshold: int = 0
    reservations: dict = field(default_factory=dict)  # order_id -> quantity

    def exists(self) -> bool:
        return bool(self.product_id)

    def available(self) -> int:
        return self.on_hand - self.reserved

    def is_low_stock(self) -> bool:
        return self.available() < self.low_stock_threshold


def rebuild_state(event_book: types.EventBook | None) -> InventoryState:
    """Rebuild inventory state from events."""
    state = InventoryState()

    if event_book is None or not event_book.pages:
        return state

    for page in event_book.pages:
        if not page.event:
            continue

        if page.event.type_url.endswith("StockInitialized"):
            event = domains.StockInitialized()
            page.event.Unpack(event)
            state.product_id = event.product_id
            state.on_hand = event.quantity
            state.low_stock_threshold = event.low_stock_threshold

        elif page.event.type_url.endswith("StockReceived"):
            event = domains.StockReceived()
            page.event.Unpack(event)
            state.on_hand = event.new_on_hand

        elif page.event.type_url.endswith("StockReserved"):
            event = domains.StockReserved()
            page.event.Unpack(event)
            state.reserved += event.quantity
            state.reservations[event.order_id] = event.quantity

        elif page.event.type_url.endswith("ReservationReleased"):
            event = domains.ReservationReleased()
            page.event.Unpack(event)
            qty = state.reservations.pop(event.order_id, 0)
            state.reserved -= qty

        elif page.event.type_url.endswith("ReservationCommitted"):
            event = domains.ReservationCommitted()
            page.event.Unpack(event)
            qty = state.reservations.pop(event.order_id, 0)
            state.on_hand -= qty
            state.reserved -= qty

    return state
